### PR TITLE
Improve spooling performance and memory accounting

### PR DESCRIPTION
**Pull request by wendigo**
## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/operator/OutputSpoolingOperatorFactory.java [hunk 1]
```diff
@@ -316,25 +315,37 @@ private static Block[] emptyBlocks(Map<Symbol, Integer> layout)
 
             return blocks;
         }
+
+        @Override
+        public void close()
+                throws Exception
+        {
+            userMemoryContext.close();
+        }
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/operator/OutputSpoolingOperatorFactory.java:319]: I do not think it needs explicit closing. At least we do not do that elsewhere

File: core/trino-main/src/main/java/io/trino/operator/OutputSpoolingOperatorFactory.java [hunk 2]
```diff
@@ -316,25 +315,37 @@ private static Block[] emptyBlocks(Map<Symbol, Integer> layout)
...
     }
 
     private static class PageBuffer
     {
-        private final List<Page> buffer;
+        private final List<Page> buffer = new ArrayList<>();
+        private final LocalMemoryContext memoryContext;
+
+        private long size;
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/operator/OutputSpoolingOperatorFactory.java:327]: no need. Use `memoryContext.getBytes`

File: testing/trino-testing/src/main/java/io/trino/testing/LocalSpoolingManager.java [hunk 3]
```diff
@@ -167,26 +173,78 @@ public SpoolingManager create(Map<String, String> config, SpoolingManagerContext
...
+        @JsonIgnore
+        public Path getPath()
+        {
+            return path;
+        }
+
+        @JsonProperty("path")
+        public String getFilePath()
+        {
+            return path.toAbsolutePath().toString();
```
Comment (losipiuk) [inline testing/trino-testing/src/main/java/io/trino/testing/LocalSpoolingManager.java:227]: why not keep it a record an just keep absolute path on field?

File: core/trino-main/src/main/java/io/trino/operator/OutputSpoolingOperatorFactory.java (hunk 1 above)
Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/operator/OutputSpoolingOperatorFactory.java:319]: It is by using closer

File: testing/trino-testing/src/main/java/io/trino/testing/LocalSpoolingManager.java (hunk 3 above)
Comment (wendigo) [reply testing/trino-testing/src/main/java/io/trino/testing/LocalSpoolingManager.java:227]: Path is serialized differently (like an URI with a file:/// scheme)

File: core/trino-main/src/main/java/io/trino/operator/OutputSpoolingOperatorFactory.java (hunk 1 above)
Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/operator/OutputSpoolingOperatorFactory.java:319]: Close is called in multiple places around the code.

Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/operator/OutputSpoolingOperatorFactory.java:319]: Actually there is a comment that I saw:

```
// caller should close this context as it's a new context
    public LocalMemoryContext newLocalUserMemoryContext(String allocationTag)
    {
        return new InternalLocalMemoryContext(operatorMemoryContext.newUserMemoryContext(allocationTag), memoryFuture, this::updatePeakMemoryReservations, true);
    }
```

**Review by losipiuk — APPROVED** [review]

**Pull request merged**